quickwit-storage/src/storage_with_upload_cache.rs

```diff
@@ -0,0 +1,1171 @@
+            .exceeds_capacity(num_bytes, num_files)
+        {
+            return Err(StorageErrorKind::InternalError.with_error(anyhow::anyhow!(
+                "Initial cache items should not exceed cache capacity.",
+            )));
+        }
+
+        let mut disk_cache = LruCache::unbounded();
+        for (path, size) in cache_state.items {
+            disk_cache.put(path, size);
+        }
+
+        Ok(Self {
+            remote_storage,
+            cache_storage,
+            cache_storage_root,
+            disk_capacity: cache_state.disk_capacity,
+            disk_cache: Mutex::new(CacheWithMeta {
+                cache: disk_cache,
+                num_bytes,
+                num_files,
+            }),
+        })
+    }
+
+    /// Persist the state of the entire cache.
+    ///
+    /// Takes a previously locked guard for easier mutex management.
+    async fn save_state(&self, wrapped_cache: &MutexGuard<'_, CacheWithMeta>) -> StorageResult<()> {
+        let mut cache_items = vec![];
+        for (path, size) in wrapped_cache.cache.iter() {
+            cache_items.push((path.clone(), *size));
+        }
+
+        let cache_state = CacheState {
+            disk_capacity: self.disk_capacity,
+            items: cache_items,
+        };
+
+        write_cache_state(self.cache_storage_root.as_path(), cache_state)
+    }
+}
+
+#[async_trait]
+impl Storage for StorageWithUploadCache {
+    async fn put(&self, path: &Path, payload: PutPayload) -> StorageResult<()> {
+        self.remote_storage.put(path, payload.clone()).await?;
+
+        let mut locked_disk_cache = self.disk_cache.lock().await;
+        let payload_length = payload.len().await? as usize;
+        // Ignore storing an item that cannot fit in the cache.
+        if payload_length > self.disk_capacity.max_num_bytes {
+            warn!("Failed to cache file: file size exceeds total cache capacity.");
+            return Ok(());
+        }
+
+        if let Some(item_num_bytes) = locked_disk_cache.cache.pop(&path.to_path_buf()) {
+            if let Err(error) = self.cache_storage.delete(path).await {
+                if error.kind() != StorageErrorKind::DoesNotExist {
+                    warn!(file_path = %path.to_string_lossy(), "Could not replace file from cache.");
+                    locked_disk_cache
+                        .cache
+                        .put(path.to_path_buf(), item_num_bytes);
+                    return Ok(());
+                }
+            }
+            locked_disk_cache.num_bytes -= item_num_bytes;
+            locked_disk_cache.num_files -= 1;
+        }
+
+        let mut undable_to_delete = HashSet::new();
+        while self.disk_capacity.exceeds_capacity(
+            locked_disk_cache.num_bytes + payload_length,
+            locked_disk_cache.num_files + 1,
+        ) {
+            if let Some((item_path, item_num_bytes)) = locked_disk_cache.cache.pop_lru() {
+                // check for cycle
```

Comment:
Can you expand on what kind of cycles can we expect and when they do happen?

Reply:
Yes. I will explain in the comment.
we can end up with files that cannot be removed remaining in the cache and creating an infinite loop.